### PR TITLE
OP-55: Self-hosted auth — argon2id, JWT, rotating refresh tokens, security.yml

### PR DESCRIPTION
**Pull request by m-nweke**
Implements OP-55 (E7): self-hosted authentication per ADR-0003, plus `security.yml`.

**Stacked on #76 (OP-54).** Base is `op-54-analyses-endpoints`; review that one first. This PR's own diff is ~2,500 lines across 19 files.

## What

Four endpoints — `POST /api/v1/auth/{register,login,refresh,logout}` — behind two new primitives in `openposture_api.security`, plus a boot guard in `config.py` and a four-job security workflow.

E6 had already laid the `users` and `refresh_tokens` tables and their repositories, so this ticket mostly fills in stubs: `auth.py` was a placeholder raising 401.

## Why the choices are what they are

**argon2id, not passlib.** Passlib's last release predates several Python versions this project supports. A password hashing library is the wrong dependency to have go quiet.

**SHA-256 for refresh tokens, argon2id for passwords — and the deciding variable is entropy of the input, not speed.** A `secrets.token_urlsafe(32)` token carries 256 bits of randomness: there is no dictionary to guess from, so a slow hash buys nothing while putting a deliberately expensive function on the path every client hits on a timer. A human password lives in a space small enough to enumerate, which is exactly what argon2's cost parameters exist to make expensive. Measured on a dev laptop: argon2id at the shipped defaults runs ~23 ms, or ~43 guesses/sec, against the ~10⁹/s a GPU manages on bare SHA-256 — roughly 28 bits of added cost. The 64 MiB `memory_cost` is what stops an attacker buying that back with parallel hardware, and it is why argon2**id** rather than bcrypt.

**Decoding names the algorithm it accepts.** A JWT's `alg` header is attacker-supplied data. A decoder that reads it can be handed `{"alg":"none"}`, or an RS256 token whose "signature" is an HMAC keyed with the public key. `algorithms=[ALGORITHM]` makes both unrepresentable rather than merely untested, and there is a test asserting the header says `none` *and* that decoding still fails.

**`require=["exp","iat","sub"]`** — the subtle one. Without it, a token carrying no `exp` passes expiry validation, because there is nothing to find expired. Absent and satisfied must not look alike.

**A failed sign-in costs the same either way.** Returning early when no user exists is a timing oracle: ~0.1 ms for a missing account against ~23 ms for a wrong password, a 200× difference far above network jitter. An attacker with a list of addresses times the responses and learns which are registered — the leak ADR-0003 forbids, arriving through the clock rather than the body. So the hash is verified regardless, against a dummy hash computed once at import. This equalises; it does not achieve constant time, and the docstring says so.

**Rotation makes theft detectable.** Every refresh retires the presented token and mints a new one in the same family. A retired token arriving twice means two parties hold it, and which one is the attacker is unknowable from the server — so the whole family is revoked and everyone re-authenticates. Revoking only the replayed row would leave whichever party refreshed most recently in possession of a working session.

**The refresh token appears in no response body.** It exists only as `HttpOnly; SameSite=Lax; Secure; Path=/api/v1/auth`. JavaScript cannot read an `HttpOnly` cookie, so the "nothing in `localStorage`" criterion is a property of the API rather than a rule the frontend has to keep remembering. `Path` scoping also keeps it off every image upload.

## Two additions beyond the ticket's literal wording

- **`min_length=32` on `jwt_secret`.** PyJWT warns that HMAC keys below RFC 7518 §3.2's floor are the weakest part of the scheme; a short key is brute-forceable *offline* from one captured token, with no request to the API and no log entry. A warning nobody reads is not a control. It immediately caught a 21-character key in `test_app_factory.py`.
- **The `_as_utc` helper.** SQLite returns naive datetimes where Postgres returns aware ones, and Python refuses to compare the two. That strictness is a feature — silently comparing a local time to a UTC one is the worse bug — but the code has to handle both backends.

## Known gaps, deliberate

- **Registration leaks whether an address is registered**, via the 409. The standard fix is to always return 201 and email the address owner; ADR-0003 records that this project has no outbound mail path, and a 201 that silently creates nothing lies to legitimate users. Documented in the route's docstring.
- **Concurrent refresh is not serialised.** Two simultaneous refreshes of the same token both read it unrevoked and both rotate, producing two live branches rather than a detected replay. Closing this needs `SELECT ... FOR UPDATE` on the token row; it is a correctness edge, not a hole an attacker can steer, and it belongs in its own ticket.
- **No rate limiting** on login or refresh. Not in this ticket's scope.
- **`get_current_user_id` remains a stub** — OP-56 owns it, and it is already in review.
- `PasswordHasher.check_needs_rehash()` (transparent upgrade when parameters are retuned) is not wired up; noted for later.

## How tested

- **Python: 306 passed**, `ruff check` + `format` clean, `mypy` clean across 46 files.
- **Frontend: 78 passed**, `tsc -b` clean, `oxlint` and prettier clean.
- Abuse cases are built as attacks, not asserted as intentions — each test forges something: expired tokens, tampered payloads, a token signed with the wrong key, `alg: none`, HS512 signed with the real secret, a valid signature over a non-UUID subject, replay, family revocation, logout invalidating a family, and production refusing to boot on the default key.
- **The frontend guard was mutation-tested.** I injected a real `localStorage.setItem("access_token", "eyJ…")` into `InMemoryAuthProvider` and confirmed all three layers — source scan, `setItem`-key scan, runtime spy — fail independently, then restored and confirmed a clean tree. A guard test that has never failed is one you are trusting on faith.
- Contract regenerated and verified idempotent; `pr.yml`'s `contract` job diffs it, so staleness is a build failure.

Pre-existing and unrelated: `packages/pose-backends/tests/test_cli.py::test_a_real_photograph_produces_a_full_table` fails identically on `main` (verified by checking out `main` and re-running). It needs the native inference stack.

## Self-review found one real bug

`_unauthorized_refresh` documented that it cleared the dead cookie, and it did not. Raising `HTTPException` hands control to the error handler, which builds its own response — so the `Set-Cookie` written to the injected `Response` was discarded. The 401 path now returns an RFC 9457 `JSONResponse` with the header attached, and two regression tests pin both the cookie and the error envelope. Fixed in 3be6844.

## Notes for the reviewer

`security.yml` cannot be verified without running it. Every action reference was checked against the real tag list rather than inferred: `actions/dependency-review-action` publishes **no major tag past v3** (`v4` and `v5` are 404s), so it is pinned to `v5.0.0`; `github/codeql-action@v4` exists and is treated as first-party under `pr.yml`'s policy; `gitleaks` is SHA-pinned as third-party and needs no licence key for a personal-account repo. The repo is public, so CodeQL and dependency review are free — both would need Advanced Security on a private one. Expect CodeQL's `security-extended` pack to surface findings on its first run; that is it working.

### CODE REVIEW

File: apps/api/tests/test_auth.py [hunk 1]
```diff
@@ -0,0 +1,457 @@
...
+        assert _without_request_id(wrong_password.json()) == _without_request_id(
+            no_such_user.json()
+        )
+
+    def test_neither_response_names_the_field_that_was_wrong(self, client: TestClient) -> None:
+        client.post(REGISTER, json=_credentials())
+
+        detail = client.post(LOGIN, json=_credentials(password="nope")).json()["detail"]
+
+        assert "email" not in detail.lower() or "password" in detail.lower()
```
[the remark this answers is not in the document]
Comment (m-nweke) [reply apps/api/tests/test_auth.py]: Valid, and it turned out to be hiding a second, worse problem. Fixed in 9442595.

**On your point:** you're right. The old assertion was an implication, `"email" not in detail or "password" in detail`, and it's satisfied by exactly the message it existed to catch:

| detail | old | new |
| --- | --- | --- |
| `Email or password is incorrect.` | PASS | PASS |
| `Password incorrect` | **PASS** | FAIL |
| `Email not found` | FAIL | FAIL |

The asymmetry is the easy thing to miss: "Email not found" obviously leaks that the address is unregistered, but "Password incorrect" leaks the *opposite* and is just as useful to an attacker — it confirms the address **is** registered, which is precisely what enumerating a list is for. Now `("email" in detail) == ("password" in detail)`, so the message must blame both or neither.

**The bigger problem your comment surfaced:** the test was vacuous. It logged in with `password="nope"` — 4 characters, under `MIN_PASSWORD_LENGTH = 12` — so the request died at schema validation and never reached the credential check at all:

```
password='nope'               -> 422  detail='The request did not match the expected schema.'
password='the-wrong-password' -> 401  detail='Email or password is incorrect.'
```

That 422 detail mentions neither field, so *any* assertion about field names would have passed. The test could not have failed for the reason it existed. It now uses a password long enough to reach the route and asserts `status_code == 401` first, so a future schema change can't quietly hollow it out again.

**Verified by mutation**, since a test that has never failed is being trusted on faith: I temporarily set `_INVALID_CREDENTIALS = "Password incorrect."`, confirmed the new assertion fails with `'password incorrect.' singles out one credential`, and restored. The old assertion passed under that same mutation.

Thanks — this was a good catch.
